// @sanity/pkg-utils/dist/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 });
var build = require("./_chunks-cjs/build.cjs"), check = require("./_chunks-cjs/check.cjs"), consoleSpy = require("./_chunks-cjs/consoleSpy.cjs"), logger = require("./_chunks-cjs/logger.cjs"), init = require("./_chunks-cjs/init.cjs"), index = require("./_chunks-cjs/index.cjs"), watch = require("./_chunks-cjs/watch.cjs");
function defineConfig(configOptions) {
  return configOptions;
}
exports.build = build.build;
exports.resolveBuildTasks = build.resolveBuildTasks;
exports.check = check.check;
exports.DEFAULT_BROWSERSLIST_QUERY = consoleSpy.DEFAULT_BROWSERSLIST_QUERY;
exports.loadConfig = consoleSpy.loadConfig;
exports.loadPkg = consoleSpy.loadPkg;
exports.loadPkgWithReporting = consoleSpy.loadPkgWithReporting;
exports.loadTSConfig = consoleSpy.loadTSConfig;
exports.parseExports = consoleSpy.parseExports;
exports.parseStrictOptions = consoleSpy.parseStrictOptions;
exports.pkgExtMap = consoleSpy.pkgExtMap;
exports.resolveConfigProperty = consoleSpy.resolveConfigProperty;
exports.strictOptions = consoleSpy.strictOptions;
exports.toggle = consoleSpy.toggle;
exports.createLogger = logger.createLogger;
exports.isRecord = logger.isRecord;
exports.createFromTemplate = init.createFromTemplate;
exports.defineTemplateOption = init.defineTemplateOption;
exports.init = init.init;
exports.buildTaskHandlers = index.buildTaskHandlers;
exports.dtsTask = index.dtsTask;
exports.dtsWatchTask = index.dtsWatchTask;
exports.getExtractMessagesConfig = index.getExtractMessagesConfig;
exports.rollupLegacyTask = index.rollupLegacyTask;
exports.rollupReactCompilerTask = index.rollupReactCompilerTask;
exports.rollupTask = index.rollupTask;
exports.rollupWatchTask = index.rollupWatchTask;
exports.watchTaskHandlers = index.watchTaskHandlers;
exports.watch = watch.watch;
exports.defineConfig = defineConfig;
